ts/javascript/prettier.ts

```diff
@@ -0,0 +1,89 @@
+import { Check } from "../check.ts";
+import { Executable } from "../executable.ts";
+import { escapeShellArg, nixRaw, nixStrLit } from "../nix.ts";
+import { Package } from "../package.ts";
+import { Plugin } from "../project.ts";
+
+const mkConfigFile = (config: Record<string, unknown>) =>
+  nixRaw`(pkgs.writeText "prettier-config" ${nixStrLit(
+    JSON.stringify(config),
+  )})`;
+
+/**
+ * A garn plugin that adds a `format` `Executable` that formats your code using
+ * prettier, and a `Check` to verify that your code is formatted.
+ *
+ * @param opts.glob An optional glob pattern to limit what to format/check
+ * @param opts.config An optional prettier configuration
+ *
+ * It will use the version of prettier defined in your package.json, otherwise
+ * if that is not configured, it will fallback to using prettier from nixpkgs.
+ *
+ * Example:
+ * ```typescript
+ * export const myProject = mkNpmProject({ ... })
+ *   .add(garn.javascript.prettier({
+ *     glob: "./src",
+ *     config: {
+ *       trailingComma: "es5",
+ *       tabWidth: 4,
+ *       semi: false,
+ *       singleQuote: true,
+ *     },
+ *   }));
+ * ```
+ *
+ * Then running `garn run myProject.format` will format formattable files under
+ * `src` with the given config. Running `garn check myProject` will error if
+ * any files under `src` need to be formatted.
+ */
+export function plugin(
+  opts: {
+    glob?: string;
+    config?: Record<string, unknown>;
```

Comment:
I couldn't find one unfortunately. The prettier repo is javascript, it has a few types but I couldn't find one for the config. `@types/prettier` is just a repo with a readme instructing you to use the types in `prettier`.

Additionally, we don't necessarily know which version of prettier the user is using. I think it might be better to rely on prettier's runtime errors to instruct the user.

Reply:
Actually nevermind, I completely missed it documented in `index.d.ts` in the installed npm package. Still not sure if this is ideal though given versioning requirements. Let me know what you think!

Reply:
Thinking about this more I think it's probably a good idea to do and we can always make it more lax in the future if there are bad version discrepancies.

Reply:
Actually it looks like deno docland does not support npm dependencies, and using the proxies like jspm don't expose the typescript types from what I can gather

Reply:
Yeah, this was just an idea. :shrug: